File: ch4/untyped_bigstep.hs
-- exercise 4.2.2: a big-step evaluator for the untyped arithmetic language

data Term = T
          | F
          | IfThenElse Term Term Term
          | Z
          | Succ Term
          | Pred Term
          | IsZero Term
          deriving Show

-- a predicate for values
isValue :: Term -> Bool
isValue T = True
isValue F = True
isValue Z = True
isValue (Succ x) = isNumeric x
isValue _ = False

-- we need this to properly enforce the "succ nv" value
isNumeric :: Term -> Bool
isNumeric Z = True
isNumeric (Succ x) = isNumeric x
isNumeric _ = False

-- big-step evaluation rules
--   we're imposing an evaluation strategy and order by how these
--   rules are framed

eval :: Term -> Term

-- B-Value
eval v
  | isValue v = v

-- B-IfTrue, B-IfFalse
eval ite@(IfThenElse b t f) = case eval b of
  T -> eval t
  F -> eval f
  _ -> ite

-- B-Succ
eval sx@(Succ x) = let x' = eval x in
  if isNumeric x' then (Succ x') else sx

-- B-PredZero, B-PredSucc
eval px@(Pred x) = case eval x of
  Z -> Z
  (Succ x)
    | isNumeric x -> x
  _ -> px

-- B-IsZeroZero, B-IsZeroSucc
eval izx@(IsZero x) = case eval x of
  Z -> T
  (Succ x)
    | isNumeric x -> F
  _ -> izx

-- either a Left stuck or a Right value
evalEither :: Term -> Either Term Term
evalEither t = let t' = eval t in
  if isValue t'
    then Right t'
    else Left t'

File: ch4/untyped_bigstep.rs
use std::rc::Rc;

#[derive(Clone, Debug)]
enum Term {
    T,
    F,
    IfThenElse(Rc<Term>, Rc<Term>, Rc<Term>),
    Z,
    Succ(Rc<Term>),
    Pred(Rc<Term>),
    IsZero(Rc<Term>),
}

impl Term {
    fn is_value(&self) -> bool {
        match *self {
            Term::T | Term::F | Term::Z => true,
            Term::Succ(ref x) => x.is_numeric_value(),
            _ => false,
        }
    }

    fn is_numeric_value(&self) -> bool {
        match *self {
            Term::Z => true,
            Term::Succ(ref x) => x.is_numeric_value(),
            _ => false,
        }
    }

    fn eval(&self) -> Term {
        match *self {
            // B-Value
            ref v if v.is_value() => v.clone(),

            // B-IfTrue & B-IfFalse
            Term::IfThenElse(ref c, ref t, ref f) => match c.eval() {
                Term::T => (**t).clone(),
                Term::F => (**f).clone(),
                _ => self.clone(),
            },

            // B-Succ
            Term::Succ(ref x) => {
                let v = x.eval();
                if v.is_numeric_value() {
                    v
                } else {
                    self.clone()
                }
            },

            Term::Pred(ref x) => match x.eval() {
                Term::Z => Term::Z,
                Term::Succ(ref x) if x.is_numeric_value() => (**x).clone(),
                _ => self.clone(),
            },

            Term::IsZero(ref x) => match x.eval() {
                Term::Z => Term::T,
                Term::Succ(ref x) if x.is_numeric_value() => Term::F,
                _ => self.clone(),
            },

            // unreachable
            _ => self.clone(),
        }
    }
}

fn main() {
    let ex = Term::IfThenElse(
        Rc::new(Term::IsZero(Rc::new(Term::Succ(Rc::new(Term::Z))))),
        Rc::new(Term::Pred(Rc::new(Term::Succ(Rc::new(Term::Z))))),
        Rc::new(Term::Succ(Rc::new(Term::Succ(Rc::new(Term::Z))))));
    println!("ex = {:?}", ex);
    println!("ex.eval() = {:?}", ex.eval());
}
